<a id='back'></a> 
# Table of Contents

* [Introduction](#intro)
<br>
<br>
* [Step 1. Downloading Libraries and Data](#step1)
    * [1.1 Import Libraries](#step1.1)
    * [1.2 Establish Connection to the Database](#step1.2)
<br>
<br>
* [Step 2. Viewing the Contents of Tables](#step2)
    * [2.1 Table `books`](#step2.1)
    * [2.2 Table `authors`](#step2.2)
    * [2.3 Table `publishers`](#step2.3)
    * [2.4 Table `ratings`](#step2.4)
    * [2.5 Table `reviews`](#step2.5)
<br>
<br>
* [Step 3. Completing the Tasks](#step3)
    * [3.1 Calculate the number of books released after January 1, 2000](#step3.1)
    * [3.2 Calculate the number of user reviews and the average rating for each book](#step3.2)
    * [3.3 Identify the publisher that has published the most books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis)](#step3.3)
    * [3.4 Identify the author with the highest average rating for books: find books with a minimum of 50 `rating_id`](#step3.4)
    * [3.5 Calculate the average number of reviews among users who have rated more than 50 books](#step3.5)

<a id='intro'></a>
## Introduction

SQL Project
<br>
The sudden appearance of the coronavirus has changed the daily routines of people worldwide. Now, city dwellers no longer spend their leisure time outside the home, such as going to cafes and malls. They are more often at home, spending their time reading books. This has prompted startup companies to develop a new application for book lovers.
<br>
<br>
You have been given a database from one of the companies competing in this industry. The database contains data about books, publishers, authors, as well as customer ratings and reviews of related books. This information will be used to create pricing offers for a new product.
<br>
Tasks

1. Calculate the number of books released after January 1, 2000.
2. Calculate the number of user reviews and the average rating for each book.
3. Identify the publisher that has published the most books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
4. Identify the author with the highest average rating for books: find books with a minimum rating of 50.
5. Calculate the average number of review texts among users who have rated more than 50 books.

<a id='step1'></a>
## Step 1. Download Libraries and Data
<a id='step1.1'></a>
### 1.1 Import Library

In [1]:
import pandas as pd
from sqlalchemy import create_engine
from sqlalchemy.engine.reflection import Inspector

### 1.2 Connecting to the database
<a id='step1.2.1'></a>

In [2]:
db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # database name

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Displays a list of tables in the database

In [3]:
# Database structure inspection
inspector = Inspector.from_engine(engine)

# Retrieve the table name
table_names = inspector.get_table_names()

# Show table name
print("Tables that exist in the database:")
for table_name in table_names:
    print(table_name)


C:\Users\Cokroaminoto\AppData\Local\Temp\ipykernel_7132\3173433516.py:2: SADeprecationWarning: The from_engine() method on Inspector is deprecated and will be removed in a future release.  Please use the sqlalchemy.inspect() function on an Engine or Connection in order to acquire an Inspector. (deprecated since: 1.4)
  inspector = Inspector.from_engine(engine)


Tables that exist in the database:
author
second
second_b
advertisment_costs
authors
books
orders
publishers
ratings
reviews
visits


The tables used are `books` , `authors` , `publishers`, `ratings` and `reviews`

[Back to Table of Contents](#back)

<a id='step2'></a>
## Step 2. View the contents of the table
<a id='step2.1'></a>
### 2.1 Table `books`

Create a SQL query for the `books` table

In [4]:
query_books = """
SELECT *
FROM books
"""

Run SQL queries using Pandas

In [5]:
books_df = pd.io.sql.read_sql(query_books, con = engine)

Displays the contents of the table

In [6]:
books_df

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

Displays general information

In [7]:
books_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


The `books_df` table contains data about books consisting of 1,000 rows, 6 columns and no missing values, with specifications
1. Column `book_id` — contains the book ID of data type 'int64'
2. Column `author_id` — contains the author ID of data type 'int64'
3. Column `title` — contains the title of the book with data type 'object'
4. Column `num_pages` — contains the number of pages of data type 'int64'
5. Column `publication_date` — contains the publication date with data type 'object'
6. Column `publisher_id` — contains the publisher ID of data type 'int64'

[Back to Table of Contents](#back)

<a id='step2.2'></a>
### 2.2 Table `authors`

Create a SQL query for the `authors` table

In [8]:
query_authors = """
SELECT *
FROM authors
"""

Run SQL queries using Pandas

In [9]:
authors_df = pd.io.sql.read_sql(query_authors, con = engine)

Displays the contents of the table

In [10]:
authors_df

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

Displays general information

In [11]:
authors_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


The `authors_df` table contains data about books consisting of 636 rows, 2 columns and no missing values, with specifications
1. Column `author_id` — contains the author ID of data type 'int64'
2. Column `author` — contains the name of the author with the data type 'object'

[Back to Table of Contents](#back)

<a id='step2.3'></a>
### 2.3 Table `publishers`

Create a SQL query for the `publishers` table

In [12]:
query_publishers = """
SELECT *
FROM publishers
"""

Run SQL queries using Pandas

In [13]:
publishers_df = pd.io.sql.read_sql(query_publishers, con = engine)

Displays the contents of the table

In [14]:
publishers_df

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

Displays general information

In [15]:
publishers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


The `publishers_df` table contains data about books consisting of 340 rows, 2 columns and no missing values, with specifications
1. Column `publisher_id` — contains the publisher ID of data type 'int64'
2. Column `publisher` — contains the name of the publisher with data type 'object'

[Back to Table of Contents](#back)

<a id='step2.4'></a>
### 2.4 Table `ratings`

Create a SQL query for the `ratings` table

In [16]:
query_ratings = """
SELECT *
FROM ratings
"""

Run SQL queries using Pandas

In [17]:
ratings_df = pd.io.sql.read_sql(query_ratings, con = engine)

Displays the contents of the table

In [18]:
ratings_df

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

Displays general information

In [19]:
ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


The `ratings_df` table contains data about books consisting of 6456 rows, 4 columns and no missing values, with specifications
1. Column `rating_id` — contains the rating ID of data type 'int64'
2. Column `book_id` — contains the book ID of data type 'int64'
3. Column `username` — contains the name of the user who rated the book with the data type 'object'
4. Column `rating` — contains ratings of data type 'int64'

[Back to Table of Contents](#back)

<a id='step2.5'></a>
### 2.5 Table `reviews`

Create a SQL query for the `reviews` table

In [20]:
query_reviews = """
SELECT *
FROM reviews
"""

Run SQL queries using Pandas

In [21]:
reviews_df = pd.io.sql.read_sql(query_reviews, con = engine)

Displays the contents of the table

In [22]:
reviews_df

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Displays general information

In [23]:
reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


The `reviews_df` table contains data about books consisting of 2793 rows, 4 columns and no missing values, with specifications
1. Column `review_id` — contains the reviewer ID of data type 'int64'
2. Column `book_id` — contains the book ID of data type 'int64'
3. Column `username` — contains the name of the user who rated the book with the data type 'object'
4. Column `text` — contains review text with data type 'int64'

[Back to Table of Contents](#back)

<a id='step3'></a>
## Step 3. Completing the Task

<a id='step3.1'></a>
### 3.1 Count the number of books released after January 1, 2000

Create a SQL query for the first task

In [24]:
query_task1 = """
SELECT COUNT(*)
FROM books
WHERE publication_date > '2000-01-01';
"""

Run SQL queries using Pandas

In [25]:
results_query_task1 = pd.io.sql.read_sql(query_task1, con = engine)

Displays the contents of `results_query_task1`

In [26]:
results_query_task1 

count
0    819

The number of books released after January 1, 2000 was 819

[Back to Table of Contents](#back)

<a id='step3.2'></a>
### 3.2 Count the number of user reviews and average rating for each book

Create a SQL query for the second task

In [27]:
query_task2 = """
SELECT 
    b.book_id,
    a.author,
    b.title,
    COUNT(DISTINCT r.review_id) AS number_users_reviews,
    AVG(ra.rating) AS average_rating,
    COUNT(ra.rating_id) AS count_rating
FROM 
    books b
JOIN 
    authors a ON b.author_id = a.author_id
LEFT JOIN 
    reviews r ON b.book_id = r.book_id
LEFT JOIN 
    ratings ra ON b.book_id = ra.book_id
GROUP BY 
    b.book_id, a.author, b.title
ORDER BY 
    number_users_reviews DESC,
    average_rating DESC
LIMIT 10;
"""

Run SQL queries using Pandas

In [28]:
results_query_task2 = pd.io.sql.read_sql(query_task2, con = engine)

Displays the contents of `results_query_task2`

In [29]:
results_query_task2

book_id                             author  \
0      948                    Stephenie Meyer   
1      302         J.K. Rowling/Mary GrandPré   
2      299         J.K. Rowling/Mary GrandPré   
3      656  Markus Zusak/Cao Xuân Việt Khương   
4      734                    Jeannette Walls   
5      497                     Diana Gabaldon   
6      750                     J.R.R. Tolkien   
7      695                        Mark Haddon   
8      779                       Rick Riordan   
9      963                         Sara Gruen   

                                               title  number_users_reviews  \
0                            Twilight (Twilight  #1)                     7   
1  Harry Potter and the Prisoner of Azkaban (Harr...                     6   
2  Harry Potter and the Chamber of Secrets (Harry...                     6   
3                                     The Book Thief                     6   
4                                   The Glass Castle                     6   
5                          Outlander (Outlander  #1)                     6   
6                The Hobbit  or There and Back Again                     6   
7  The Curious Incident of the Dog in the Night-Time                     6   
8  The Lightning Thief (Percy Jackson and the Oly...                     6   
9                                Water for Elephants                     6   

   average_rating  count_rating  
0        3.662500          1120  
1        4.414634           492  
2        4.287500           480  
3        4.264151           318  
4        4.206897           174  
5        4.125000           144  
6        4.125000           528  
7        4.081081           222  
8        4.080645           372  
9        3.977273           264

Displays the book title without being cut off

In [30]:
pd.set_option('display.max_colwidth', None)

Displays the contents of the dataframe

In [31]:
results_query_task2

book_id                             author  \
0      948                    Stephenie Meyer   
1      302         J.K. Rowling/Mary GrandPré   
2      299         J.K. Rowling/Mary GrandPré   
3      656  Markus Zusak/Cao Xuân Việt Khương   
4      734                    Jeannette Walls   
5      497                     Diana Gabaldon   
6      750                     J.R.R. Tolkien   
7      695                        Mark Haddon   
8      779                       Rick Riordan   
9      963                         Sara Gruen   

                                                         title  \
0                                      Twilight (Twilight  #1)   
1  Harry Potter and the Prisoner of Azkaban (Harry Potter  #3)   
2   Harry Potter and the Chamber of Secrets (Harry Potter  #2)   
3                                               The Book Thief   
4                                             The Glass Castle   
5                                    Outlander (Outlander  #1)   
6                          The Hobbit  or There and Back Again   
7            The Curious Incident of the Dog in the Night-Time   
8    The Lightning Thief (Percy Jackson and the Olympians  #1)   
9                                          Water for Elephants   

   number_users_reviews  average_rating  count_rating  
0                     7        3.662500          1120  
1                     6        4.414634           492  
2                     6        4.287500           480  
3                     6        4.264151           318  
4                     6        4.206897           174  
5                     6        4.125000           144  
6                     6        4.125000           528  
7                     6        4.081081           222  
8                     6        4.080645           372  
9                     6        3.977273           264

From the query results above, we can see the 10 books with the highest number of user reviews, sorted by the number of user reviews (number_users_reviews) in descending order. The following are the conclusions from these results:

1. **Twilight (Twilight #1) by Stephenie Meyer**: Has the highest number of user reviews, namely 7 reviews, with an average rating of 3.6625.
2. **Harry Potter and the Prisoner of Azkaban (Harry Potter #3) by J.K. Rowling/Mary GrandPre**: Ranked second with 6 user reviews and an average rating of 4.414634.
3. **Harry Potter and the Chamber of Secrets (Harry Potter #2) by J.K. Rowling/Mary GrandPré**: As with the previous book, also has 6 user reviews, with a slightly lower average rating of 4.2875.
4. **The Book Thief by Markus Zusak/Cao Xuân Việt Khương**: Has 6 user reviews and an average rating of 4.264151.
5. **The Glass Castle by Jeannette Walls**: Also has 6 user reviews with an average rating of 4.206897.
6. **Outlander (Outlander #1) by Diana Gabaldon**: As before, has 6 user reviews with an average rating of 4,125.
7. **The Hobbit or There and Back Again by J.R.R. Tolkien**: Also has 6 user reviews with an average rating of 4,125.
8. **The Curious Incident of the Dog in the Night-Time by Mark Haddon**: Has 6 user reviews with an average rating of 4.081081.
9. **The Lightning Thief (Percy Jackson and the Olympians #1) by Rick Riordan**: Also has 6 user reviews with an average rating of 4.080645.
10. **Water for Elephants by Sara Gruen**: Same as before, has 6 user reviews with an average rating of 3.977273.

In conclusion, the Twilight book (Twilight #1) by Stephenie Meyer is ranked first with the highest number of user reviews.

If the number_users_reviews value is the same for multiple books in the query results, then the order between those books may be determined randomly, depending on how the database system organizes its default order.

Reset the display

In [32]:
pd.set_option('display.max_colwidth', 100)

[Back to Table of Contents](#back)

<a id='step3.3'></a>
### 3.3 Identify publishers that have published the largest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis)

Create a SQL query for the third task

In [33]:
query_task3 = """
SELECT
    p.publisher_id,
    p.publisher,
    COUNT(b.book_id) AS total_books_published
FROM
    publishers p
JOIN
    books b ON p.publisher_id = b.publisher_id
WHERE
    b.num_pages > 50
GROUP BY
    p.publisher_id, p.publisher
ORDER BY
    COUNT(b.book_id) DESC
LIMIT 10;
"""

Run SQL queries using Pandas

In [34]:
results_query_task3 = pd.io.sql.read_sql(query_task3, con = engine)

Displays the contents of `results_query_task3`

In [35]:
results_query_task3

publisher_id                 publisher  total_books_published
0           212             Penguin Books                     42
1           309                   Vintage                     31
2           116  Grand Central Publishing                     25
3           217          Penguin Classics                     24
4            33          Ballantine Books                     19
5            35                    Bantam                     19
6            45                   Berkley                     17
7           284        St. Martin's Press                     14
8            46             Berkley Books                     14
9            83                     Delta                     13

From the query results above, we can see the top 10 publishers based on the number of books published with more than 50 pages. The following are the conclusions from these results:

1. **Penguin Books**: Is the top publisher with the most books published (42 books).
2. **Vintage**: Second top publisher with 31 books.
3. **Grand Central Publishing**: Placed third with 25 books.
4. **Penguin Classics**: Comes in fourth with 24 books.
5. **Ballantine Books** and **Bantam**: Both have the same number of books, namely 19 books, so they are in fifth place.
6. **Berkley** and **Berkley Books**: Both have the same number of books, namely 17 books, so they are in sixth place.
7. **St. Martin's Press**: Has 14 books.
8. **Delta**: Is in ninth place with 13 books.

In conclusion, Penguin Books is the largest publisher with the largest number of books published in this category.

[Back to Table of Contents](#back)

<a id='step3.4'></a>
### 3.4 Identify the author with the highest average book rating: find books with a `rating_id` count of at least 50

Create a SQL query for the fourth task

In [36]:
query_task4 = """
SELECT
    a.author_id,
    a.author,
    COUNT(r.rating_id) AS count_rating,
    AVG(r.rating) AS average_rating,
    COUNT(DISTINCT r.username) AS count_reviewers
FROM
    authors a
JOIN
    books b ON a.author_id = b.author_id
JOIN
    ratings r ON b.book_id = r.book_id
GROUP BY
    a.author_id, a.author
HAVING
    COUNT(r.rating_id) >= 50
ORDER BY
    average_rating DESC;
"""

Run SQL queries using Pandas

In [37]:
results_query_task4 = pd.io.sql.read_sql(query_task4, con = engine)

Displays the contents of `results_query_task4`

In [38]:
results_query_task4

author_id                                              author  \
0         130                                      Diana Gabaldon   
1         236                          J.K. Rowling/Mary GrandPré   
2           3                                     Agatha Christie   
3         402                   Markus Zusak/Cao Xuân Việt Khương   
4         240                                      J.R.R. Tolkien   
5         499                            Roald Dahl/Quentin Blake   
6         376                                   Louisa May Alcott   
7         498                                        Rick Riordan   
8          39                                       Arthur Golden   
9         542                                        Stephen King   
10        302                                        John Grisham   
11        621                                     William Golding   
12        445                                     Nicholas Sparks   
13        291                                        Jodi Picoult   
14        536                                     Sophie Kinsella   
15        257                                     James Patterson   
16        235                                       J.D. Salinger   
17        469            Paulo Coelho/Alan R. Clarke/Özdemir İnce   
18        630  William Shakespeare/Paul Werstine/Barbara A. Mowat   
19        106                                           Dan Brown   
20        372                                          Lois Lowry   
21        195             George Orwell/Boris Grabnar/Peter Škerl   
22        554                                     Stephenie Meyer   
23        311                                      John Steinbeck   

    count_rating  average_rating  count_reviewers  
0             50        4.300000               45  
1            312        4.288462              152  
2             53        4.283019               45  
3             53        4.264151               53  
4            166        4.240964              122  
5             62        4.209677               52  
6             54        4.203704               53  
7             84        4.130952               74  
8             56        4.107143               54  
9            106        4.009434               75  
10            70        3.971429               58  
11            71        3.901408               71  
12           111        3.882883               81  
13            76        3.881579               70  
14            57        3.877193               52  
15            64        3.859375               54  
16            98        3.846939               93  
17            57        3.789474               57  
18            66        3.787879               66  
19           143        3.741259              107  
20            65        3.738462               62  
21            74        3.729730               74  
22           160        3.662500              160  
23            73        3.643836               70

From the results of this query, there are 24 authors who have a minimum rating of 50 with the conclusion:

1. Diana Gabaldon has the highest average rating of authors with a minimum of 50 ratings, with an average rating of 4.3.
2. J.K. Rowling/Mary GrandPré ranked second with an average rating of 4,288, but with a much higher number of ratings at 312, indicating great popularity.
3. The majority of authors have an average rating above 4, which indicates that their books are generally liked by readers.
4. Some famous writers such as J.R.R. Tolkien, Roald Dahl, and Agatha Christie are also on the list with high average ratings.
5. Stephenie Meyer, author of Twilight, has a large number of ratings (160), but her average rating is somewhat lower than the other authors on this list, namely 3,662.

Membuat kueri SQL untuk tugas keempat

In [39]:
query_task4 = """
SELECT
    a.author_id,
    a.author,
    COUNT(r.rating_id) AS count_rating,
    AVG(r.rating) AS average_rating,
    COUNT(DISTINCT r.username) AS count_reviewers
FROM
    authors a
JOIN
    books b ON a.author_id = b.author_id
JOIN
    ratings r ON b.book_id = r.book_id
GROUP BY
    a.author_id, a.author
HAVING
    COUNT(r.rating_id) >= 50
ORDER BY
    average_rating DESC;
"""

Menjalankan kueri SQL menggunakan Pandas

In [40]:
results_query_task4 = pd.io.sql.read_sql(query_task4, con = engine)

Menampilkan isi `results_query_task4`

In [41]:
results_query_task4

author_id                                              author  \
0         130                                      Diana Gabaldon   
1         236                          J.K. Rowling/Mary GrandPré   
2           3                                     Agatha Christie   
3         402                   Markus Zusak/Cao Xuân Việt Khương   
4         240                                      J.R.R. Tolkien   
5         499                            Roald Dahl/Quentin Blake   
6         376                                   Louisa May Alcott   
7         498                                        Rick Riordan   
8          39                                       Arthur Golden   
9         542                                        Stephen King   
10        302                                        John Grisham   
11        621                                     William Golding   
12        445                                     Nicholas Sparks   
13        291                                        Jodi Picoult   
14        536                                     Sophie Kinsella   
15        257                                     James Patterson   
16        235                                       J.D. Salinger   
17        469            Paulo Coelho/Alan R. Clarke/Özdemir İnce   
18        630  William Shakespeare/Paul Werstine/Barbara A. Mowat   
19        106                                           Dan Brown   
20        372                                          Lois Lowry   
21        195             George Orwell/Boris Grabnar/Peter Škerl   
22        554                                     Stephenie Meyer   
23        311                                      John Steinbeck   

    count_rating  average_rating  count_reviewers  
0             50        4.300000               45  
1            312        4.288462              152  
2             53        4.283019               45  
3             53        4.264151               53  
4            166        4.240964              122  
5             62        4.209677               52  
6             54        4.203704               53  
7             84        4.130952               74  
8             56        4.107143               54  
9            106        4.009434               75  
10            70        3.971429               58  
11            71        3.901408               71  
12           111        3.882883               81  
13            76        3.881579               70  
14            57        3.877193               52  
15            64        3.859375               54  
16            98        3.846939               93  
17            57        3.789474               57  
18            66        3.787879               66  
19           143        3.741259              107  
20            65        3.738462               62  
21            74        3.729730               74  
22           160        3.662500              160  
23            73        3.643836               70

Dari hasil query tersebut, ada 24 penulis yang memiliki minimal jumlah rating 50 dengan kesimpulan:

1. Diana Gabaldon memiliki rata-rata rating tertinggi dari penulis yang memiliki minimal 50 rating, dengan rata-rata rating sebesar 4.3.
2. J.K. Rowling/Mary GrandPré menempati peringkat kedua dengan rata-rata rating 4.288, tetapi dengan jumlah rating yang jauh lebih tinggi yaitu 312, menunjukkan popularitas yang besar.
3. Mayoritas penulis memiliki rata-rata rating di atas 4, yang menunjukkan bahwa buku-buku mereka umumnya disukai oleh pembaca.
4. Beberapa penulis terkenal seperti J.R.R. Tolkien, Roald Dahl, dan Agatha Christie juga masuk dalam daftar dengan rata-rata rating yang tinggi.
5. Stephenie Meyer, penulis Twilight, memiliki jumlah rating yang besar (160), tetapi rata-rata ratingnya agak lebih rendah dibandingkan dengan penulis lain di daftar ini, yaitu 3.662.

[Back to Table of Contents](#back)

<a id='step3.5'></a>
### 3.5 Calculate the average number of review texts among users who rated more than 50 books

Create a SQL query for the fifth task

In [42]:
query_task5 = """
SELECT
    AVG(subquery2.review_cnt) AS avg_review_cnt,
    AVG(subquery2.text_length) AS avg_text_review_length
FROM
    (SELECT
        COUNT(reviews.review_id) AS review_cnt,
        subquery1.username,
        AVG(LENGTH(reviews.text)) AS text_length
    FROM
        (SELECT
            ra.username,
            COUNT(ra.rating_id) AS rating_cnt
        FROM
            ratings ra
        GROUP BY
            ra.username
        HAVING
            COUNT(ra.rating_id) > 50) AS subquery1
    LEFT JOIN reviews ON reviews.username = subquery1.username
    GROUP BY
        subquery1.username) AS subquery2;

"""

Run SQL queries using Pandas

In [43]:
results_query_task5 = pd.io.sql.read_sql(query_task5, con = engine)

Displays the contents of `results_query_task5`

In [44]:
results_query_task5

avg_review_cnt  avg_text_review_length
0       24.333333               87.934632

After querying, the average user who rated more than 50 books created 24.33 reviews with an average character count of 88 characters.

[Back to Table of Contents](#back)